Symplectic Maximals

TristanPfersdorff opened the pull request:
- Add `MaximalSubgroupClassRepsSpecialUnitaryGroup`
- Modify `ClassicalMaximalsGeneric`
- Various fixes for classes C2, C5, C6, C7
- Test `MaximalSubgroupClassRepsSpecialUnitaryGroup`

With this PR, the maximal geometric subgroups of the symplectic groups should be all done.

## Checklist for the reviewer
### General
- [ ] All functions have unit tests

### Functions constructing generators of maximal subgroups
- [ ] The code which computes and then stores the sizes is correct. To do this one confers to the tables referenced by the code. Tests for the group size should use `RECOG.TestGroup` from the recog package.
- [ ] The test suite checks whether all constructed generators are sensible. That is it tests the generators'
  - [ ] determinants (and if applicable also the spinor norms are correct), and
  - [ ] compatibility with the correct form,
  - [ ] `DefaultFieldOfMatrixGroup` returns the correct field.

### Functions assembling the list of all maximal subgroups of a certain group
- [ ] The code agrees with the tables in section 8.2 of the book "The Maximal Subgroups of the Low-dimensional Finite Classical Groups".

The reviewer doesn't need to compare our results to magma's results. That's the job of the person implementing the code.

fingolfin commented on gap/ClassicalMaximals.gi:
```diff
@@ -997,18 +1015,135 @@ function(n, q)
+        # Class C1 subgroups ######################################################
+        # Cf. Propositions 3.3.1 (n = 4), 3.5.1 (n = 6), 3.7.1 (n = 8),
+        #                  3.9.1 (n = 10), 3.11.1 (n = 12) in [BHR13]
+        maximalSubgroups := Concatenation(maximalSubgroups,
+                                          C1SubgroupsSymplecticGroupGeneric(n, q));
```
```suggestion
        Append(maximalSubgroups, C1SubgroupsSymplecticGroupGeneric(n, q));
```

fingolfin commented on gap/ClassicalMaximals.gi:
```diff
@@ -997,18 +1015,135 @@ function(n, q)
+        #                  3.7.3, 3.7.4 (all n = 8), 3.9.3, 3.9.4 (all n = 10),
+        #                  3.11.3, 3.11.4, 3.11.5, 3.11.6 (all n = 12) in [BHR13]
+        if not (n = 4 and q = 3) then
+            maximalSubgroups := Concatenation(maximalSubgroups,
+                                              C2SubgroupsSymplecticGroupGeneric(n, q));
```
```suggestion
            Append(maximalSubgroups, C2SubgroupsSymplecticGroupGeneric(n, q));
```

fingolfin commented on gap/ClassicalMaximals.gi:
```diff
@@ -997,18 +1015,135 @@ function(n, q)
+        # Cf. Propositions 3.3.4 (n = 4), 3.5.5 (n = 6), 3.7.5 (n = 8),
+        #                  3.9.5 (n = 10), 3.11.7 (n = 12) in [BHR13]
+        if not (n = 4 and q = 3) then
+            maximalSubgroups := Concatenation(maximalSubgroups,
+                                          C3SubgroupsSymplecticGroupGeneric(n, q));
```
```suggestion
            Append(maximalSubgroups, C3SubgroupsSymplecticGroupGeneric(n, q));
```

fingolfin replied:
and so on...

fingolfin commented on gap/ClassicalMaximals.gi:
```diff
@@ -846,13 +848,12 @@ end);
-                            List(listOfks, 
+    result := List([1..QuoInt(n, 2)], k -> SpStabilizerOfIsotropicSubspace(n, q, k));
+    # type Sp(2 * k, q) _|_ Sp(n - 2 * k, q) subgroups
+    result := Concatenation(result,
```
auch hier: Append...

fingolfin commented on gap/ClassicalMaximals.gi:
```diff
@@ -942,26 +944,42 @@ end);
+
+        # Cf. Proposition 4.5.4 (i) in [KL90] for the number of conjugates
+        numberOfConjugates := Gcd(2, q - 1, r);
+        result := Concatenation(result, ConjugatesInGeneralGroup(G, generatorNormSpMinusSp, numberOfConjugates));
```
Append

fingolfin commented on gap/ClassicalMaximals.gi:
```diff
@@ -997,18 +1015,135 @@ function(n, q)
+
+    if Length(classes) = 0 then
+        classes := [1..9];
+    elif IsList(classes[1]) then
```
```suggestion
    elif Length(classes) = 1 and IsList(classes[1]) then
```
So that `MaximalSubgroupClassRepsSymplecticGroup(n, q, [1,2,4], 5)` is not accepted

fingolfin commented on gap/ReducibleMatrixGroups.gi:
```diff
@@ -391,13 +391,13 @@ end);
     fi;
 
-    if k >= d / 2 then
+    if k > d / 2 then
         ErrorNoReturn("<k> must be less than <d> / 2");
```
The error message does not fit the condition anymore, does it?

fingolfin reviewed (commented):
Some minor comments, otherwise looks good to me.

TristanPfersdorff commented:
Alright, changes are implemented. While I was at it, I fixed all instances of silly `x := Concatenation(x, ...)` syntax that I could conceivably find with a keen eye and Ctrl + f (about 50 of them, wtf we're we doing?!) and replaced it by `Append(x, ...)`. I also added the safer checks to disallow stuff like `MaximalSubgroupClassRepsSymplecticGroup(n, q, [1,2,4], 5)` to some other functions since those checks are done in a few more functions inside `ClassicalMaximals.gi`.

TristanPfersdorff commented:
As @maxhauck pointed out to me, not all of the constructed groups had been conjugated to the standard GAP forms after construction, which would cause issues if GAP ever changed its standard form. I hope to have taken care of that in the last commit, though some of the construction functions do not explicitly set the form before conjugating. Presumably, this is because we do not know the form in those cases. That issue should be largely fixed by #88 though.
